source/frontend/editors/RarcEditor.cpp

```diff
@@ -0,0 +1,1035 @@
+
+    if (size_tier == 0) {
+      size_string = std::format("{} {}", node_size, suffix);
+    } else {
+      size_string = std::format(
+          "{} {}",
+          static_cast<std::size_t>(node_size / std::pow(1000, size_tier)),
+          suffix);
+    }
+  }
+
+  ImGui::Text("%s", size_string.c_str());
+
+  ImGui::PopID();
+}
+
+void RarcEditorPropertyGrid::DrawOperationModal(RarcEditor* editor) {
+  if (m_operation_state == TriState::ST_INDETERMINATE)
+    return;
+
+  std::string dialog_name =
+      m_operation_state == TriState::ST_FALSE ? "Failure" : "Success";
+
+  if (m_operation_modal_state == ModalState::M_OPENING) {
+    ImGui::OpenPopup(dialog_name.c_str());
+  }
+
+  auto flags = ImGuiWindowFlags_NoCollapse | ImGuiWindowFlags_NoMove |
+               ImGuiWindowFlags_AlwaysAutoResize;
+  bool open = true;
+
+  if (ImGui::BeginPopupModal(dialog_name.c_str(), &open, flags)) {
+    RSL_DEFER(ImGui::EndPopup());
+
+    m_operation_modal_state = ModalState::M_OPEN;
+
+    ImGui::Text(m_operation_state == TriState::ST_FALSE
+                    ? "Operation failed!"
+                    : "Operation completed successfully!");
+
+    if (!open) {
+      ImGui::CloseCurrentPopup();
+      m_operation_modal_state = ModalState::M_CLOSED;
+      return;
+    }
+
+    auto* enter_key = ImGui::GetKeyData(ImGuiKey_Enter);
+    if (enter_key->Down) {
+      ImGui::CloseCurrentPopup();
+      m_operation_modal_state = ModalState::M_CLOSED;
+    }
+  }
+}
+
+static void
+get_sorted_directory_list_r(const std::filesystem::path& path,
+                            std::vector<std::filesystem::path>& out) {
+  std::vector<std::filesystem::path> dirs;
+  for (auto&& it : std::filesystem::directory_iterator{path}) {
+    if (it.is_directory())
+      dirs.push_back(it.path());
+    else
+      out.push_back(it.path());
+  }
+  for (auto& dir : dirs) {
+    out.push_back(dir);
+    get_sorted_directory_list_r(dir, out);
+  }
+}
+
+static void renameNode(ResourceArchive& rarc,
+                       std::optional<ResourceArchive::Node>& node,
+                       const std::string& new_name) {
+  RSL_DEFER(node = std::nullopt);
+
+  if (!node)
+    return;
+
+  if (new_name == "")
+    return;
+
+  auto node_it = std::find(rarc.nodes.begin(), rarc.nodes.end(), *node);
+  if (node_it == rarc.nodes.end())
+    return;
+
+  node_it->name = new_name;
+}
+
+static bool insertFiles(ResourceArchive& rarc,
+                        std::optional<ResourceArchive::Node> parent,
+                        std::vector<rsl::File>& files) {
+  RSL_DEFER(files.clear());
+
+  if (files.size() == 0)
+    return false;
+
+  if (!parent)
+    return false;
+
+  std::vector<ResourceArchive::Node> new_nodes;
+  for (auto& file : files) {
+    auto file_name = file.path.filename().string();
+    std::transform(file_name.begin(), file_name.end(), file_name.begin(),
+                   [](u8 ch) { return std::tolower(ch); });
```

Comment:
We use this a lot, maybe this would be better as like `rsl::to_lower`?

Reply:
Yeah that would be better.